### PR TITLE
Update OpenBSD bcrypt algorithm

### LINKED ISSUE #54071: outdated bcrypt algorithm in use
##### SUMMARY
Older bcrypt implementation is in use

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
encrypt

##### ANSIBLE VERSION

```paste below
2.7+
```

##### CONFIGURATION

```paste below
Not applicable
```

##### OS / ENVIRONMENT

Discovered via [OpenBSD patch](https://github.com/openbsd/ports/blob/master/sysutils/ansible/patches/patch-lib_ansible_utils_encrypt_py)

##### STEPS TO REPRODUCE

```yaml
Not applicable
```

##### EXPECTED RESULTS

Update to useing `$2b$` for the modern bcrypt usage.

##### ACTUAL RESULTS

```paste below
Not applicable
```

### PR DESCRIPTION
**Pull request by ghost**
##### SUMMARY

Adds a conditional check for OpenBSD to use $2b for the bcrypt algorithm instead of $2a. Its a workaround for the issue linked below until a better fix is possible. This eliminates the need for a custom patch on that platform.

Fixes: #54071

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

encrypt library

##### ADDITIONAL INFORMATION

Downstream patch located here: https://github.com/openbsd/ports/blob/master/sysutils/ansible/patches/patch-lib_ansible_utils_encrypt_py

```paste below

```

### CODE REVIEW

File: lib/ansible/utils/encrypt.py
Comment (Akasurde) [inline lib/ansible/utils/encrypt.py]:
```suggestion
    system = to_text(platform.system(), errors='surrogate_or_strict')
    if system == u"OpenBSD":
```

**Review by Akasurde — DISMISSED** [review]
Could you please add integration test for this change in [filters](https://github.com/ansible/ansible/blob/devel/test/integration/targets/filters/tasks/main.yml)?

**Comment by evitalis** [comment]
@Akasurde I took a stab at adding the tests, let me know what you think or if more changes are needed. I also merged in your proposed changes to the `encrypt.py` file.

File: lib/ansible/utils/encrypt.py
Comment (ghost) [reply lib/ansible/utils/encrypt.py]: Added this change in. Thanks for the input.

**Comment by evitalis** [comment]
@Akasurde Just wanted to follow up on this and see if my implemented test met the expectation or if I needed to make other changes?

**Comment by evitalis** [comment]
I not longer use OpenBSD current maintainer there is @jasperla who should be able to fix this per the OpenBSD patching policy.

**Comment by evitalis** [comment]
@jasperla Any input on this as it will save you dealing with additional, unnecessary, patched for your port.

**Comment by Akasurde** [comment]
@mattclay Could you please advise us on test failures ? Thanks.

**Comment by mattclay** [comment]
The test needs to install `bcrypt` before it can be used. That should be done here:

https://github.com/ansible/ansible/blob/6adf0c581ee005f514ea68e80c0f91e97e63dcea/test/integration/targets/filters/runme.sh#L10

**Comment by Akasurde** [comment]
@webknjaz Can you please help me with error in CI ? Thanks in advance.

**Comment by mattclay** [comment]
@webknjaz This is the failure @Akasurde is referring to:

https://app.shippable.com/github/ansible/ansible/runs/133737/64/console

```
06:59 Traceback (most recent call last):
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/cli/base_command.py", line 188, in main
06:59     status = self.run(options, args)
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/commands/install.py", line 345, in run
06:59     resolver.resolve(requirement_set)
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/legacy_resolve.py", line 196, in resolve
06:59     self._resolve_one(requirement_set, req)
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/legacy_resolve.py", line 359, in _resolve_one
06:59     abstract_dist = self._get_abstract_dist_for(req_to_install)
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/legacy_resolve.py", line 307, in _get_abstract_dist_for
06:59     self.require_hashes
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/operations/prepare.py", line 215, in prepare_linked_requirement
06:59     finder, self.build_isolation,
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/distributions/source.py", line 70, in prepare_distribution_metadata
06:59     reqs = self.req.pep517_backend.get_requires_for_build_wheel()
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 71, in get_requires_for_build_wheel
06:59     'config_settings': config_settings
06:59   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 162, in _call_hook
06:59     raise BackendUnavailable
06:59 BackendUnavailable
```

**Comment by webknjaz** [comment]
@mattclay ah.. so my patch didn't help. System site-packages "leak" into an isolated build env. Can we run `pip install --no-use-pep517 bcrypt` before running the whole thing? Alternatively, upgrading OS-level setuptools would fix this.

Or maybe we could change Pip invocation globally to `pip install --no-build-isolation`?

**Comment by webknjaz** [comment]
I've changed it to `--no-build-isolation`, let's see what does CI have to say about it...

File: lib/ansible/utils/encrypt.py
Comment (webknjaz) [inline lib/ansible/utils/encrypt.py]: I'd probably prefer a single assignment with a conditional only for the argument. It would probably look cleaner this way:
```suggestion
    is_openbsd = u"OpenBSD" == to_text(platform.system(), errors='surrogate_or_strict')
    bcrypt_crypt_id = '2b' if is_openbsd else '2a'
    algorithms['bcrypt'] = algo(crypt_id=bcrypt_crypt_id, salt_size=22, implicit_rounds=None)
```

File: examples/hosts
Comment (webknjaz) [inline examples/hosts]: @Akasurde @evitalis could you please revert this part of the patch? `git checkout 43d6fdc~1 -- examples/hosts && git commit examples/hosts` should fix it, but this needs to be done locally, I cannot fix it via GitHub UI.

File: test/integration/targets/filters/runme.sh
Comment (webknjaz) [inline test/integration/targets/filters/runme.sh]: @Akasurde @mattclay this fixes the issue. Keep it here until Pip fixes its bug.

Comment (mattclay) [reply test/integration/targets/filters/runme.sh]: @webknjaz Thanks. Can you add a comment explaining why the option is needed and when it can be removed? If there's a bug to link to, I'd include that in the comment.

File: test/integration/targets/filters/tasks/main.yml
Comment (webknjaz) [inline test/integration/targets/filters/tasks/main.yml]: Let's have more Unicode tests
```suggestion
      - "'{{ 'café' | password_hash(hashtype='blowfish', salt='1234567890123456789123') }}' == '$2b$12$123456789012345678912uDZsaKh/j7qZkKps5QEOGP4UfpxNa56S'"
      - "'{{ 'Криївка' | password_hash(hashtype='blowfish', salt='1234567890123456789123') }}' == '$2b$12$123456789012345678912utpRGnLLKUmQLWXHzHrg8U2Lwths9TNu'"
```

**Review by webknjaz — APPROVED** [review]

File: test/integration/targets/filters/runme.sh
Comment (webknjaz) [reply test/integration/targets/filters/runme.sh]: I'm thinking https://github.com/pypa/pip/issues/6264#issuecomment-480100770 but nobody has verified it to be 100% sure.

**Comment by evitalis** [comment]
@webknjaz Did you still need me to remove that blank line from the file or has the issue been resolved since it looks like the code now has 1 approval.

**Comment by webknjaz** [comment]
I still want that removed. It's totally unrelated to the issue. I've added my approval tho because I'm happy with the main part of the patch.

**Comment by evitalis** [comment]
@webknjaz The example hosts file should now be removed. I think it slipped in from a different PR I had opened, sorry about that.

**Review by webknjaz — APPROVED** [review]

**Comment by webknjaz** [comment]
This PR may need rebasing.

File: test/integration/targets/filters/tasks/main.yml
Comment (webknjaz) [inline test/integration/targets/filters/tasks/main.yml:378]: > 07:03     "assertion": "'$2a$12$123456789012345678912unE0GeDfx0eAsIsscM2jcQtjpIY2LVuy' == '$2b$12$123456789012345678912unE0GeDfx0eAsIsscM2jcQtjpIY2LVuy'", 

Ubuntu 16.04 -> https://app.shippable.com/github/ansible/ansible/runs/136075/72/console

> 07:02 ERROR: Exception:
07:02 Traceback (most recent call last):
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/cli/base_command.py", line 188, in main
07:02     status = self.run(options, args)
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/commands/install.py", line 345, in run
07:02     resolver.resolve(requirement_set)
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/legacy_resolve.py", line 196, in resolve
07:02     self._resolve_one(requirement_set, req)
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/legacy_resolve.py", line 359, in _resolve_one
07:02     abstract_dist = self._get_abstract_dist_for(req_to_install)
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/legacy_resolve.py", line 307, in _get_abstract_dist_for
07:02     self.require_hashes
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/operations/prepare.py", line 215, in prepare_linked_requirement
07:02     finder, self.build_isolation,
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_internal/distributions/source.py", line 70, in prepare_distribution_metadata
07:02     reqs = self.req.pep517_backend.get_requires_for_build_wheel()
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 71, in get_requires_for_build_wheel
07:02     'config_settings': config_settings
07:02   File "/root/ansible_testing/venv/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 162, in _call_hook
07:02     raise BackendUnavailable
07:02 BackendUnavailable

FreeBSD 11.1 -> https://app.shippable.com/github/ansible/ansible/runs/136075/64/console

**Review by webknjaz — CHANGES_REQUESTED** [review]
Turns out there's still CI failures

**Comment by evitalis** [comment]
@webknjaz Does anything else need done from my side here? Perhaps that test simply needs limited to the affected platform? Have not seen an update in a bit and I imagine the ansible code has had some changes in the past month.

Did @jasperla ever weigh in as well on additional testing?

**Comment by Akasurde** [comment]
@evitalis Could you please work on the failing test cases in this PR ? I was not able to give time to fix the failing test case.

**Comment by evitalis** [comment]
@Akasurde @webknjaz Have not had time to work on this, nor do I have an OpenBSD system to test, but did have a thought. Is it possible the Jinja2 filter is not properly using the bcrypt module on Linux? As far as I understand it, blowfish isn't part of glibc on Linux which might explain some of the issue.

Are all hashes being changed to use the '2b' prefix or just those for OpenBSD? It looks like the latter but the test is trying to compare Linux hashes to those with '2b' prefixes when they may potentially end up with '2a'. I might be misreading the newest changes though.

**Comment by evitalis** [comment]
Per the ansible core team meeting today I tried to move this logic around and remove the conditional. I couldn't run the integration tests locally and the CI looks to have issues.

Per #54071 the prefix change should be part of only the PasslibHash section. Not too sure how to properly implement this.

Would it be possible to get a hand to get me in the right direction please? @webknjaz

File: lib/ansible/utils/encrypt.py
Comment (s-hertel) [inline lib/ansible/utils/encrypt.py:63]: I don't think there's a need to change this.

File: lib/ansible/utils/encrypt.py
Comment (s-hertel) [inline lib/ansible/utils/encrypt.py]: Named tuples are immutable, so you can't reset one of the items. You'll need to set the whole tuple.

In addition, self.algorithms is a class attribute of the base class rather than an instance attribute. So instead, perhaps update the class attribute right before the `__init__` of the PasslibHash class: 
```
class PasslibHash(BaseHash):

    BaseHash.algorithms['bcrypt'] = BaseHash.algo(crypt_id='2b', salt_size=22, implicit_rounds=None)

    def __init__(self, algorithm):
        ...
```

I didn't run all the failing tests, but that resolved the integration test for lookups that are failing.

File: lib/ansible/utils/encrypt.py
Comment (ghost) [reply lib/ansible/utils/encrypt.py:63]: The original PR did not change this but I can pull it out for my next commit. I am not sure why this was added. Though it might be a py2-ism.

File: lib/ansible/utils/encrypt.py
Comment (ghost) [reply lib/ansible/utils/encrypt.py]: Thanks. I will update this for the next commit and see what happens. I do appreciate the help @s-hertel a little out of my element on this one.

Comment (s-hertel) [reply lib/ansible/utils/encrypt.py]: no problem, thanks for continuing to work on this PR

**Comment by evitalis** [comment]
@webknjaz @s-hertel I noticed something odd this morning while updating this PR. I see the error about the mismatched hashes in shippable, but locally I cannot reproduce it. If I manually install bcrypt and passlib, both without `--build-isolation`, I get a hash with `2b` since passlib switched it to being the default.

Adding @sivel due to #54071. The change was made in v1.7.0 [(changelog)](https://passlib.readthedocs.io/en/stable/history/1.7.html#id7) so this PR is likely going about solving the issue wrong anyway given that information. The change is listed under the "other changes" section and was done in 2016.

I am not sure the right way to rework this given the information of the change of default and as far as I can tell it was never changed back to `2a`.

**Comment by evitalis** [comment]
While I would like to see the newer implementation of bcrypt used I am withdrawing this PR. There appears to be a lack of interest from @jasperla or the Ansible team in getting the code up to modern defaults across the platforms, and I have other priorities at the moment.

**Pull request closed without merging**
